pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go

```diff
@@ -729,48 +739,50 @@ func (lrw *logicalReplicationWriterProcessor) flushBuffer(
 		}
 	}
 
-	minChunk := 64
+	minChunk, maxChunk := minChunkSize.Default(), maxChunkSize.Default()
 	if lrw.FlowCtx != nil {
-		minChunk = int(minChunkSize.Get(&lrw.FlowCtx.Cfg.Settings.SV))
+		minChunk = minChunkSize.Get(&lrw.FlowCtx.Cfg.Settings.SV)
+		maxChunk = maxChunkSize.Get(&lrw.FlowCtx.Cfg.Settings.SV)
 	}
+	chunks := make(chan []streampb.StreamEvent_KV)
 
-	chunkSize := max((len(kvs)/len(lrw.bh))+1, minChunk)
-
-	perChunkStats := make([]flushStats, len(lrw.bh))
+	// Aim for a chunk size that gives each worker at least 10 chunks to do, but
+	// with a minimum chunk size and maximum target as well.
+	chunkSize := min(max(len(kvs)/(len(lrw.bh)*10), int(minChunk)), int(maxChunk))
+	requiredWorkers := max(1, min(len(kvs)/chunkSize), len(lrw.bh))
 
-	todo := kvs
 	g := ctxgroup.WithContext(ctx)
-	for worker := range lrw.bh {
-		if len(todo) == 0 {
-			perChunkStats = perChunkStats[:worker]
-			break
-		}
-		// The chunk should end after the first new key after chunk size.
-		chunkEnd := min(chunkSize, len(todo))
-		for chunkEnd < len(todo) && k(todo[chunkEnd-1]).Equal(k(todo[chunkEnd])) {
-			chunkEnd++
-		}
-		chunk := todo[0:chunkEnd]
-		todo = todo[len(chunk):]
-		bh := lrw.bh[worker]
-
-		if err := ctx.Err(); err != nil {
-			// Bail early if ctx is canceled. NB: we break rather than return the err
-			// now since we still need to Wait() to avoid leaking a goroutine. We will
-			// re-check for any ctx errors after the Wait() in case all workers had
-			// completed without error as of this break.
-			break
-		}
-
+	for worker := range lrw.bh[:requiredWorkers] {
+		w := worker
```

Comment:
no need for this local var in iterators anymore, i think

Reply:
this was supposed to be a different PR anyway.